File: _submissions/19UCS159_Tharun/Day4/minStack.java
import java.util.*;

public class minStack {
  Stack<int[]> stack = new Stack<>();

  public void push(int x) {
    int min = stack.isEmpty() ? x : Math.min(x, stack.peek()[1]);
    stack.push(new int[] { x, min });
  }

  public void pop() {
    if (stack.isEmpty())
      return;
    stack.pop();
  }

  public int top() {
    if (stack.isEmpty())
      return -1;
    return stack.peek()[0];
  }

  public int getMin() {
    if (stack.isEmpty())
      return -1;
    return stack.peek()[1];
  }

}

File: _submissions/18UEI056_Prashun/Day4/min_stack.cpp
int minElm=0;
stack<int>s;
MinStack::MinStack() {
    while(!s.empty())
    s.pop();
}

void MinStack::push(int x) {
if(s.empty())
 {s.push(x);
     minElm=x;
 }
 else{
      if(minElm<=x)
      s.push(x);
      else if(minElm>x)
      { s.push(2*x-minElm);
      minElm=x;
      }
 }
}

void MinStack::pop() {
    if(!s.empty())
    { if(minElm <=s.top())
        s.pop();
        else if(minElm>s.top())
        { minElm=2*minElm-s.top();
            s.pop();
        }
    }
    return ;
}

int MinStack::top() {
    if(!s.empty())
    { if(s.top()>=minElm)
     return s.top();
     else if(s.top()<minElm)
     return minElm;
    }
    return -1;
}

int MinStack::getMin() {
    if(s.size()==0)
    return -1;
    
     return minElm;
}

